### PR TITLE
[WIP] Parquet integration

### PR DESCRIPTION
**Pull request by msaisumanth**

### CODE REVIEW

File: ludwig/data/parquet_dataset.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        self.data_parquet_fp = data_parquet_fp
+        self.parquet_file_reader = pq.ParquetFile(self.data_parquet_fp).reader
+
+        self.size = self.parquet_file_reader.scan_contents()
+        self.features = {}
+
+        for feature in input_features + output_features:
+            self.features[feature['name']] = feature
+
+    def get(self, feature_name, idx=None):
```
Comment (ydudin3) [inline ludwig/data/parquet_dataset.py]: Nit: if idx is a list call it  idxs.

File: ludwig/data/parquet_dataset.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import pyarrow.parquet as pq
+
+
+class ParquetDataset:
+    def __init__(self, dataset, input_features, output_features, data_parquet_fp):
```
Comment (ydudin3) [inline ludwig/data/parquet_dataset.py]: Nit: what is the data type of the dataset here, can we add something to the name to make it more clear?

File: ludwig/data/parquet_dataset.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                                          self.features[feature_name]['level'])
+
+        col_idx = self.parquet_file_reader.column_name_idx(feature_name)
+
+        return self.parquet_file_reader.read_column(col_idx)[min_idx:max_idx].to_pylist()
+
+    def get_dataset(self):
+        return self.dataset
+
+    def set_dataset(self, dataset):
```
Comment (ydudin3) [inline ludwig/data/parquet_dataset.py]: Not sure we even need this, I think having self.dataset in the constructor is all we need.

File: ludwig/train_preprocessed.py [hunk 4]
```diff
@@ -0,0 +1,531 @@
+#! /usr/bin/env python
```
Comment (ydudin3) [inline ludwig/train_preprocessed.py]: Are these changes unrelated, need to rebase?

**Comment by clasnake** [comment]
@msaisumanth @ydudin3 @w4nderlust Is there any update on this PR? The feature to support parquet is heavily needed in our work.

**Comment by w4nderlust** [comment]
> @msaisumanth @ydudin3 @w4nderlust Is there any update on this PR? The feature to support parquet is heavily needed in our work.

It is something we are still working on. If it is very needed for you work, would you consider helping out on it? I can give precise instructions on how to complete the task, it has not been completed yet because there are other features we prioritized more, but what needs to be done is pretty clear.

**Comment by clasnake** [comment]
> > @msaisumanth @ydudin3 @w4nderlust Is there any update on this PR? The feature to support parquet is heavily needed in our work.
> 
> It is something we are still working on. If it is very needed for you work, would you consider helping out on it? I can give precise instructions on how to complete the task, it has not been completed yet because there are other features we prioritized more, but what needs to be done is pretty clear.

@w4nderlust That would be great. Please give me any instructions needed and I can contribute. Also we started to use Ludwig in our projects recently and it looks very promising and we'd love to use/learn/contribute more from/to it!

**Comment by w4nderlust** [comment]
> @w4nderlust That would be great. Please give me any instructions needed and I can contribute. Also we started to use Ludwig in our projects recently and it looks very promising and we'd love to use/learn/contribute more from/to it!

Sounds great! I'm curious to hear about your usecases (also a private email would be good).
Regarding how to contribute, what needs to happen before being able to add arbitrary new data formats like Parquet is the modifications proposed in this branch: https://github.com/uber/ludwig/tree/preprocessing_strategy
That code does not work yet, it's a proof of concept of how preprocessing will look like after refactoring, so first i will have to make that work.
Once that is done, I'll post here and will expain in detail how it will be possible to add Parquet processing. Does this sound good to you?

**Comment by clasnake** [comment]
> > @w4nderlust That would be great. Please give me any instructions needed and I can contribute. Also we started to use Ludwig in our projects recently and it looks very promising and we'd love to use/learn/contribute more from/to it!
> 
> Sounds great! I'm curious to hear about your usecases (also a private email would be good).
> Regarding how to contribute, what needs to happen before being able to add arbitrary new data formats like Parquet is the modifications proposed in this branch: https://github.com/uber/ludwig/tree/preprocessing_strategy
> That code does not work yet, it's a proof of concept of how preprocessing will look like after refactoring, so first i will have to make that work.
> Once that is done, I'll post here and will expain in detail how it will be possible to add Parquet processing. Does this sound good to you?

Sounds good to me! Will drop you a private mail later about our use case and we can discuss more later in detail.

**Comment by clasnake** [comment]
@w4nderlust Any update on the preprocessing strategy stuff based on which we'll be able to easily add other external data sources?

**Comment by w4nderlust** [comment]
> @w4nderlust Any update on the preprocessing strategy stuff based on which we'll be able to easily add other external data sources?

It's ongoing work, we will focus more on it after we are done with the TF2 porting. Hold tight :)

**Pull request closed without merging**